api/status_types.go

```diff
@@ -0,0 +1,345 @@
+// Copyright 2023 Red Hat, Inc. and/or its affiliates
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Based on https://github.com/knative/pkg/blob/980a33719a10024f45e44320316c5bd35cef18d6/apis/condition_set.go
+
+// Status ...
+// +kubebuilder:object:generate=true
+type Status struct {
+	// The latest available observations of a resource's current state.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions Conditions `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// The generation observed by the deployment controller.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
```

Comment:
see the meta/v1 spec. That `ObservedGeneration` is for each condition AFAIU

Reply:
Oh sure, depends on the implementation. On Knative they use it in the whole status since every reconciliation cycle they reevaluate every condition. We are following a similar approach. So it makes sense keeping in a higher level.

Also in the Deployment Status, same thing:

```yaml
status:
  observedGeneration: 3
  replicas: 1
  updatedReplicas: 1
  readyReplicas: 1
  availableReplicas: 1
  conditions:
    - type: Available
      status: 'True'
      lastUpdateTime: '2023-02-28T11:25:36Z'
      lastTransitionTime: '2023-02-28T11:25:36Z'
      reason: MinimumReplicasAvailable
      message: Deployment has minimum availability.
    - type: Progressing
      status: 'True'
      lastUpdateTime: '2023-02-28T11:55:42Z'
      lastTransitionTime: '2023-02-28T11:25:32Z'
      reason: NewReplicaSetAvailable
      message: ReplicaSet "greeting-57f9548ff4" has successfully progressed.
```